// node-collection-json/lib/Link.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _CollectionError = require('./CollectionError');

var _CollectionError2 = _interopRequireDefault(_CollectionError);

var _Entity2 = require('./Entity');

var _Entity3 = _interopRequireDefault(_Entity2);

var _Library = require('./Library');

var _Library2 = _interopRequireDefault(_Library);

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Creates a valid collection+json link object
 *
 * @author S. Fleming <nuxnik@int5.net>
 * @since Tue Jul 25 11:07:32 CEST 2017
 *
 */
var Link = function (_Entity) {
  _inherits(Link, _Entity);

  _createClass(Link, null, [{
    key: 'getByObject',

    /**
     * Get link object by json data object
     *
     * @todo - finish this
     * @param {Object} json The JSON object
     * @param {Object} config The axios configuration object. See axios documentation for more options
     * @param {Object} cache The caching object
     * @return Link
     */
    value: function getByObject(json) {
      var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var cache = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;

      //check the href
      var hrefString = Link.getObjectValueByKey(json, "href");
      if (hrefString === undefined) {
        throw new _CollectionError2.default('link.href String undefined');
      }

      //check the rel
      var relString = Link.getObjectValueByKey(json, "rel");
      if (relString === undefined) {
        throw new _CollectionError2.default('link.rel String undefined');
      }

      //check the render
      var renderString = Link.getObjectValueByKey(json, "render");

      var link = new Link(hrefString, relString, renderString, config, cache);

      //check the prompt
      var promptString = Link.getObjectValueByKey(json, "prompt");
      if (promptString !== undefined) {
        link.setPrompt(promptString);
      }

      return link;
    }

    /**
     * The class constructor
     *
     * @param string href The link uri
     * @param string rel The relational element
     * @param string render The render type
     * @param {Object} config The axios configuration object. See axios documentation for more options
     * @param {Object} cache The caching object
     */

  }]);

  function Link(href, rel) {
    var render = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
    var config = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {};
    var cache = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : null;

    _classCallCheck(this, Link);

    /**
     * The global axios client configuration object
     */
    var _this = _possibleConstructorReturn(this, (Link.__proto__ || Object.getPrototypeOf(Link)).call(this));

    _this.config = config;

    /**
     * The cache object
     *
     * @var array
     */
    _this.cache = cache;

    _this.setHref(href);
    _this.setRel(rel);
    _this.setRender(render);
    _this.setPrompt('');
    return _this;
  }

  /**
   * Get the link string
   *
   * @return string
   */


  _createClass(Link, [{
    key: 'getHref',
    value: function getHref() {
      return this.href;
    }

    /**
     * Set the link string
     *
     * @param string link The link uri
     * @return Link
     */

  }, {
    key: 'setHref',
    value: function setHref(href) {
      this.href = href;

      return this;
    }

    /**
     * Get the rel string
     *
     * @return string
     */

  }, {
    key: 'getRel',
    value: function getRel() {
      return this.rel;
    }

    /**
     * Set the rel string
     *
     * @param string rel The rel element
     * @return rel
     */

  }, {
    key: 'setRel',
    value: function setRel(rel) {
      this.rel = rel;

      return this;
    }

    /**
     * Get the link string
     *
     * @return string
     */

  }, {
    key: 'getPrompt',
    value: function getPrompt() {
      return this.prompt;
    }

    /**
     * Set the link string
     *
     * @param string link The link uri
     * @return Link
     */

  }, {
    key: 'setPrompt',
    value: function setPrompt(prompt) {
      this.prompt = prompt;

      return this;
    }

    /**
     * Get the render string
     *
     * @return string
     */

  }, {
    key: 'getRender',
    value: function getRender() {
      return this.render;
    }

    /**
     * Set the render string
     *
     * @param string render The render type string
     * @return Link
     */

  }, {
    key: 'setRender',
    value: function setRender(render) {
      this.render = render;

      return this;
    }

    /**
     * Follow a link and return the collection object
     *
     * @param {Array} params Extra params to add to the url
     * @param {Object} config The axios configuration object. See axios documentation for more options
     * @return Collection
     */

  }, {
    key: 'follow',
    value: function follow() {
      var _this2 = this;

      var params = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
      var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      // get the config values
      var mergedConfig = _Library2.default.mergeConfigurationValues(this.config, config);

      var url = this.getHref();
      if (params !== null) {
        url += '?';
        for (var key in params) {
          url += '&' + key + '=' + params[key];
        }
      }
      if (this.cache !== null && this.cache.isResourceCached(url)) {
        return Promise.resolve(this.cache.getCollectionByResource(resource));
      } else {
        return _axios2.default.get(url, mergedConfig).then(function (response) {
          response.data.collection.href = url;
          var collection = Collection.getByObject(response.data, mergedConfig, _this2.cache);
          _this2.cache.addCollection(collection);
          return collection;
        }).catch(function (error) {
          var collection = Collection.getByObject(error.response.data, _this2.config, _this2.cache);
          return collection;
        });
      }
    }

    /**
     * Get compiled json object
     *
     * @return Object
     */

  }, {
    key: 'getJson',
    value: function getJson() {
      var link = {};
      if (this.href) {
        link.href = this.getHref();
      }
      if (this.rel) {
        link.rel = this.getRel();
      }
      if (this.render) {
        link.render = this.getRender();
      }
      if (this.prompt) {
        link.prompt = this.getPrompt();
      }

      return link;
    }

    /**
     * Add collection to cache
     *
     * @param Collection collection The collection to cache
     * @return Link
     */

  }, {
    key: 'addCache',
    value: function addCache(collection) {
      if (this.cache !== null) {
        this.cache.addCollection(collection);
      }
      return this;
    }
  }]);

  return Link;
}(_Entity3.default);

exports.default = Link;
